internal/handlers/common/update.go

```diff
@@ -635,6 +714,71 @@ func extractValueFromUpdateOperator(op string, update *types.Document) (*types.D
 	return doc, nil
 }
 
+// validateRenameExpression validates $rename input on correctness.
+func validateRenameExpression(update *types.Document) error {
+	if !update.Has("$rename") {
+		return nil
+	}
+
+	updateExpression := must.NotFail(update.Get("$rename"))
+
+	doc, ok := updateExpression.(*types.Document)
+	if !ok {
+		return NewWriteErrorMsg(ErrFailedToParse, "Modifiers operate on fields but we found another type instead")
+	}
+
+	iter := doc.Iterator()
+	keys := map[string]struct{}{}
+
+	for {
+		k, v, err := iter.Next()
+		if err != nil {
+			if errors.Is(err, iterator.ErrIteratorDone) {
+				break
+			}
+
+			return err
+		}
+
+		var vStr string
+
+		vStr, ok = v.(string)
+		if !ok {
+			return NewWriteErrorMsg(
+				ErrBadValue,
+				fmt.Sprintf("The 'to' field for $rename must be a string: %s: %v", k, vStr),
+			)
+		}
+
+		if k == vStr {
+			return NewWriteErrorMsg(
+				ErrBadValue,
+				fmt.Sprintf("The source and target field for $rename must differ: %s: %v", k, vStr),
+			)
+		}
+
+		if _, ok = keys[k]; ok {
+			return NewWriteErrorMsg(
+				ErrConflictingUpdateOperators,
+				fmt.Sprintf("Updating the '%s' would create a conflict at '%s'", k, k),
+			)
+		}
+
+		keys[k] = struct{}{}
+
+		if _, ok = keys[vStr]; ok {
+			return NewWriteErrorMsg(
```

Comment:
Can we come up with a test case that would cover this error?

Reply:
Good idea! Done https://github.com/FerretDB/FerretDB/pull/1753/files#diff-02beb83969c6feb8b48d15aebcd299647e0aa2e8b1e5c531e74eb2c4f636ad85R533-R536